File: apps/haskell/design-pattern/test/CompositeTest.hs
module CompositeTest (tests) where

import Test.HUnit
import Composite

tests :: Test
tests = TestLabel "Composite" $ TestList
  [ testLeafTask
  , testCompositeTask
  , testNestedComposite
  , testAddSubTask
  ]

testLeafTask :: Test
testLeafTask = TestCase $ do
  let task = LeafTask "コーディング" 4.0 1
  assertEqual "名前" "コーディング" (taskName task)
  assertEqual "所要時間" 4.0 (timeRequired task)
  assertEqual "サブタスクなし" [] (subTasks task)
  assertEqual "優先度" 1 (priority task)

testCompositeTask :: Test
testCompositeTask = TestCase $ do
  let t1 = LeafTask "設計" 2.0 1
      t2 = LeafTask "実装" 4.0 2
      t3 = LeafTask "テスト" 3.0 1
      project = CompositeTask "プロジェクト" [t1, t2, t3] 1
  assertEqual "合計時間" 9.0 (timeRequired project)
  assertEqual "サブタスク数" 3 (length (subTasks project))

testNestedComposite :: Test
testNestedComposite = TestCase $ do
  let t1 = LeafTask "DB設計" 2.0 1
      t2 = LeafTask "API設計" 3.0 2
      backend = CompositeTask "バックエンド" [t1, t2] 1
      t3 = LeafTask "UI設計" 2.0 1
      frontend = CompositeTask "フロントエンド" [t3] 2
      project = CompositeTask "全体" [backend, frontend] 1
  assertEqual "合計時間" 7.0 (timeRequired project)

testAddSubTask :: Test
testAddSubTask = TestCase $ do
  let project = CompositeTask "プロジェクト" [] 1
      task = LeafTask "タスク1" 2.0 1
      updated = addSubTask task project
  assertEqual "サブタスク追加" 1 (length (subTasks updated))
  assertEqual "合計時間" 2.0 (timeRequired updated)

File: apps/scala/design-pattern/test/CompositeTest.scala
package designpattern.composite

import Task.*

class CompositeSuite extends munit.FunSuite:

  test("Leaf タスクの所要時間を取得する") {
    val task = Task.Leaf("コーディング", 4.0)
    assertEqualsDouble(task.getTimeRequired, 4.0, 0.01)
  }

  test("Composite タスクの所要時間は子タスクの合計") {
    val composite = Task.Composite(
      "プロジェクト",
      List(
        Task.Leaf("設計", 2.0),
        Task.Leaf("実装", 5.0),
        Task.Leaf("テスト", 3.0)
      )
    )
    assertEqualsDouble(composite.getTimeRequired, 10.0, 0.01)
  }

  test("ネストされた Composite の所要時間") {
    val subProject = Task.Composite(
      "バックエンド",
      List(
        Task.Leaf("API 設計", 2.0),
        Task.Leaf("API 実装", 8.0)
      )
    )
    val project = Task.Composite(
      "プロジェクト全体",
      List(
        subProject,
        Task.Leaf("フロントエンド", 6.0)
      )
    )
    assertEqualsDouble(project.getTimeRequired, 16.0, 0.01)
  }

  test("基本タスクの総数を取得する") {
    val project = Task.Composite(
      "プロジェクト",
      List(
        Task.Composite(
          "設計",
          List(
            Task.Leaf("要件定義", 1.0),
            Task.Leaf("アーキテクチャ", 2.0)
          )
        ),
        Task.Leaf("実装", 5.0)
      )
    )
    assertEquals(project.totalBasicTasks, 3)
  }

  test("addChild でサブタスクを追加する") {
    val project = Task.Composite("プロジェクト", List.empty)
    val updated = project.addChild(Task.Leaf("タスク1", 1.0))
    assertEquals(updated.totalBasicTasks, 1)
  }

  test("Leaf に addChild すると例外が発生する") {
    val leaf = Task.Leaf("タスク", 1.0)
    intercept[IllegalArgumentException] {
      leaf.addChild(Task.Leaf("子", 0.5))
    }
  }
